### PR TITLE
Remove RawSlice and RawSliceMut usage in butterflies.rs

### PR DESCRIPTION
**Pull request by groscoe2**
Addresses issue #98 by eliminating the usage of RawSlice and RawSliceMut in /algorithm without introducing code duplication.

I created a new internal trait LoadStore and a new internal struct DoubleBuff. LoadStore contains functions load and store, mirroring the behavior of RawSlice and RawSliceMut. DoubleBuff contains an input and output slice, mirroring structure of RawSlice and RawSliceMut being combined into one object.

All the perform_fft_contiguous functions take a single buffer value that implement LoadStore.

Normal slices, arrays, and DoubleBuff all implement LoadStore.

### CODE REVIEW

File: src/algorithm/butterflies.rs [hunk 1]
```diff
@@ -253,19 +283,18 @@ impl<T: FftNum> Butterfly4<T> {
         }
     }
     #[inline(always)]
-    pub(crate) unsafe fn perform_fft_contiguous(
+    pub(crate) unsafe fn perform_fft_butterfly<U: LoadStore<Complex<T>> + ?Sized>(
```
Comment (ejmahler) [inline src/algorithm/butterflies.rs:286]: I see this "+?sized" bound peppered everywhere. Skimming through, it doesn't look like there's anywhere that we *don't* add this - is there any way to bake this into the LoadStore trait definition?

File: src/algorithm/butterflies.rs [hunk 2]
```diff
@@ -1100,28 +1115,24 @@ impl<T: FftNum> Butterfly13<T> {
...
     #[inline(never)]
-    unsafe fn perform_fft_contiguous(
-        &self,
-        input: RawSlice<Complex<T>>,
-        output: RawSliceMut<Complex<T>>,
-    ) {
+    unsafe fn perform_fft_butterfly<U: LoadStore<Complex<T>> + ?Sized>(&self, buffer: &mut U) {
         // This function was derived in the same manner as the butterflies for length 3, 5 and 7.
         // However, instead of doing it by hand the actual code is autogenerated
         // with the `genbutterflies.py` script in the `tools` directory.
```
Comment (ejmahler) [inline src/algorithm/butterflies.rs:1121]: These bigger prime-sized routines are generated by a script, mentioned here in this comment. One thing missing from this PR is updating that script to generate correct implementations in the future.

File: src/algorithm/butterflies.rs [hunk 3]
```diff
@@ -3,20 +3,54 @@ use num_complex::Complex;
...
 
+pub(crate) trait LoadStore<T> {
+    unsafe fn load(&self, idx: usize) -> T;
+    unsafe fn store(&mut self, val: T, idx: usize);
+}
+
+impl<T: Copy> LoadStore<T> for [T] {
+    #[inline(always)]
+    unsafe fn load(&self, idx: usize) -> T {
+        *self.get_unchecked(idx)
```
Comment (ejmahler) [inline src/algorithm/butterflies.rs:18]: One thing the RawSlice struct had that these trait imps don't, is a debug_assert!() that the index is in range. I think they play an important part of maintaining peace of mind that unit tests will catch any out-of-bounds errors before they make it into a release. Could you re-add them?

**Comment by ejmahler** [comment]
I left some comments. Overall I think this is great. Thanks for putting in the work to make this change. It's elegant, it's clearly more sound than the previous implementation, and it doesn't impose any extra boilerplate on butterfly implementations. I only wish I had thought of doing it this way.

**Comment by ejmahler** [comment]
ooooo looks like we fail the MSRV 1.37 test.

I don't think that's an issue, our policy allows us to upgrade as long as it's not an upgrade to a very recent version. But we will need to decide what to upgrade to, and update the CI config, before we can mrge this.

**Comment by ejmahler** [comment]
Looks like the initial rust const generics release was 1.51, so we oculd upgrade it to that. Alternatively, we could say this is a good time to unify things, and upgrade everything to the rust version that neon requires, which is iirc rustc 1.63.

File: src/algorithm/butterflies.rs (hunk 1 above)
Comment (groscoe2) [reply src/algorithm/butterflies.rs]: I tried adding it to add it originally, but it turns out that you can't have any ?Trait in a supertrait. I was able to resolve the issue by switching to use "U" from "&mut U", adding "DerefMut" as a bound on "U", and making LoadStore<T> into a supertrait where U: LoadStoreFns<Complex<T>> + DerefMut.

File: src/algorithm/butterflies.rs (hunk 3 above)
Comment (groscoe2) [reply src/algorithm/butterflies.rs]: Certainly! I had incorrectly assumed that get_unchecked and get_unchecked_mut already had debug_asserts. Apparently I remembered the proposal for doing so without remembering that it wasn't approved!

File: src/algorithm/butterflies.rs (hunk 2 above)
Comment (groscoe2) [reply src/algorithm/butterflies.rs:1080]: I took a look at that script and it is currently generating code that was out of date before my PR (i.e. `*buffer.get_unchecked`). I've updated it and believe it is producing the correct output.

**Comment by groscoe2** [comment]
> I left some comments. Overall I think this is great. Thanks for putting in the work to make this change. It's elegant, it's clearly more sound than the previous implementation, and it doesn't impose any extra boilerplate on butterfly implementations. I only wish I had thought of doing it this way.

Thank you for the comments! I believe I have addressed them all with the latest commits.

**Comment by groscoe2** [comment]
> Looks like the initial rust const generics release was 1.51, so we oculd upgrade it to that. Alternatively, we could say this is a good time to unify things, and upgrade everything to the rust version that neon requires, which is iirc rustc 1.63.

I'm good with either of those options! We could also try to work around the const generics usage since there should be a reasonably straightforward way to coerce the arrays into mutable slices instead.

**Comment by groscoe2** [comment]
> Looks like the initial rust const generics release was 1.51, so we oculd upgrade it to that. Alternatively, we could say this is a good time to unify things, and upgrade everything to the rust version that neon requires, which is iirc rustc 1.63.

I'm good with either of those options! We could also try to work around the const generics usage since there should be a reasonably straightforward way to coerce the arrays into mutable slices instead.

**Comment by HEnquist** [comment]
Before modifying the CI, could we get this merged? https://github.com/ejmahler/RustFFT/pull/101
It's using the unmaintained actions-rs now, which use the deprecated node 12. Don't know exactly when this will finally stop working, but it will happen.

**Comment by WalterSmuts** [comment]
Why is `RawSlice` still used? I see it's used in the `avx32_butterflies` modules too. 

Gave a quick stab  at checking if miri also detects UB for those tests but run looks like miri doesn't support avx instructions:
```
thread 'avx::avx64_butterflies::unit_tests::test_avx_butterfly5_f64' panicked at 'Can't run test because this machine doesn't have the required instruction sets: ()', src/avx/avx64_butterflies.rs:1565:5
```

File: src/algorithm/butterflies.rs (hunk 2 above)
Comment (ejmahler) [reply src/algorithm/butterflies.rs:1080]: Oops lol. Unfortunately we don't have any automation making sure they're up to date. Thanks for updating it.

**Comment by ejmahler** [comment]
> Gave a quick stab at checking if miri also detects UB for those tests but run looks like miri doesn't support avx instructions:
> 
> ```
> thread 'avx::avx64_butterflies::unit_tests::test_avx_butterfly5_f64' panicked at 'Can't run test because this machine doesn't have the required instruction sets: ()', src/avx/avx64_butterflies.rs:1565:5
> ```

This error message isn't too useful heh

**Comment by ejmahler** [comment]
I upgraded the CI to test for 1.61

I'm working locally to switch avx and sse etc over to use the DoubleBuff struct, to hopefully completely eliminate the unsoundness. Once that's in I'll submit that to this PR and merge.

**Comment by ejmahler** [comment]
AVX, SSE, and Neon have now been converted, and I deleted the RawSlice struct.

There's a little too much copy/pasting of the SseArray and NeonArray trait imps for my taste. I tried making a macro for them, but it seems very hard to make a macro that borrows a different field of a struct, per macro invocation(Ie the impls for `&mut [Complex<T>]` just want to call `as_ptr()` on `self`, but the impls for `&mut DoubleBuff<'_, T>` want to call `as_ptr()` on `self.input`, and it's hard to write a macr othat lets you select one or the other per invocation).

But it's not worth holding up the PR to fix that.

> I'm good with either of those options! We could also try to work around the const generics usage since there should be a reasonably straightforward way to coerce the arrays into mutable slices instead.

Separately from this PR, I've been waiting for const generics to replace the `array_utils::iter_chunks` with the standard library's `array_chunks::<N>`, I'm hoping it will result in the same codegen with less unsafe.

**Comment by groscoe2** [comment]
> AVX, SSE, and Neon have now been converted, and I deleted the RawSlice struct.
> 
> There's a little too much copy/pasting of the SseArray and NeonArray trait imps for my taste. I tried making a macro for them, but it seems very hard to make a macro that borrows a different field of a struct, per macro invocation(Ie the impls for `&mut [Complex<T>]` just want to call `as_ptr()` on `self`, but the impls for `&mut DoubleBuff<'_, T>` want to call `as_ptr()` on `self.input`, and it's hard to write a macr othat lets you select one or the other per invocation).
> 
> But it's not worth holding up the PR to fix that.
> 
> > I'm good with either of those options! We could also try to work around the const generics usage since there should be a reasonably straightforward way to coerce the arrays into mutable slices instead.
> 
> Separately from this PR, I've been waiting for const generics to replace the `array_utils::iter_chunks` with the standard library's `array_chunks::<N>`, I'm hoping it will result in the same codegen with less unsafe.

Thank you for converting AVX, SSE, and Neon and thank you for merging the PR!

Regarding the SseArray and NeonArray trait impls, I put together PR #115 which does function pass-through implementation for DoubleBuf to remove some of the duplication. It also renames DoubleBuff to DoubleBuf to match typical Rust nomenclature and adds Deref/DerefMut to the AVX, SSE, and Neon traits to ensure that the buffers are modified in-place. This did require changing "buffer: &mut [Complex<A>]" to "mut buffer: &mut [Complex<A>]" in a couple of places. I am not super happy about this, but it does match the syntax of "mut buffer: impl LoadStore<T>", which I believe we decided is better than "buffer: &mut (impl LoadStore<T> + ?Sized)".

**Pull request merged**